Split Keyboard RGB Matrix Support

XScorpion2 opened the pull request:
## Description

So this is a bit of the ground work to get RGB Matrix split working with Split Common similar to how CRKBD works.  There are still some things to debate, so here is an overview of the changes:

- Split common now also syncs master half matrix to slave, this allows slave to know the state of the keyboard
- Split common slave now calls matrix_scan_quantum(), this allows the slave to update it's state correctly the same way master does
    - This also prevents needing to add one off support constantly for things like encoder updating or oled drivers on the slave
- timer.c is now sync'd between halves
- Small changes to RGB Matrix to know what indexes apply to which half.
- Small changes to Zygomorph Keyboard as I needed something to test this on.

Still to do & need input from others:
- [ ] Since the slave now processes keycodes, certain keycodes or systems will need checks to not run on the slave side. For example: RESET should not reset the slave as well as the master by default. I'm not sure what other keycodes or systems will need checks added for slave side, so taking suggestions.
- [x] There is a possibility the halves can get out of sync, so I want to see how folks get out of sync to know what the minimum to sync across to keep it operating correctly without just syncing everything (rgb matrix is big...)
- [x] i2c code to sync the both halves of the matrix

Known Issues:
- LT(layer, kc) with KC_CAPS causes reactive animation stutter. This is because it has an inline wait_ms(200). Not much can be done to improve this situation as it hard locks all processing of the keyboard.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
Posting more for reference. 

Works on the Corne when converted to split common (eg with #6001).
However, this breaks "normal" split keyboards. Serial is unstable, and i2c outright doesn't work.

drashna commented on quantum/quantum.c:
```diff
@@ -307,6 +307,9 @@ bool process_record_quantum(keyrecord_t *record) {
       if (record->event.pressed) {
+        #ifdef SPLIT_KEYBOARD
+        if (is_keyboard_master())
+        #endif
```
Just a fun fact, you can do this:
```c
        #ifdef SPLIT_KEYBOARD
        if (is_keyboard_master())
        #endif
        {
             reset_keyboard();
        
```

XScorpion2 commented:
Tested this quite extensively over the last month on my Zygomorph, and outside of the power related delocks, I ran into no issues.

This does appear to cause power related delocks during usb enumeration, or rather more likely to cause that. IE: If the rgb power draw was too high on the zygo during usb enumeration it would fail to fully connect and deadlock. Dropping the rgb values down slightly allowed it to connect, then I was able to put them back up to the higher values. Could use someone with more usb enumeration / hardware state debugging experience than I have to track that down.

XScorpion2 commented:
Updated the i2c split code as it was wrong. Just need @drashna to wake up and test it.

drashna commented on tmk_core/common/keyboard.c:
```diff
@@ -274,7 +283,7 @@ void keyboard_task(void)
 #endif
 
-    if (is_keyboard_master()) {
+    //if (is_keyboard_master()) {
```
This has bugged me for a while ....

I'm not sure how the corne code handles it without disabling it, but that may be a better solution than this. 

But in the meanwhile, it may be a good idea to have this as a configurable option.  Eg 
```suggestion
    #ifndef SPLIT_KEYBOARD_MIRRORING
    if (is_keyboard_master()) 
    #endif
    {
```

XScorpion2 replied:
The corne basically calls straight into what that calls bypassing it if I remember correctly.
That or is_keyboard_master() just always returns true as it's not implemented on the corne's split code.

XScorpion2 replied:
Will add this once testing is complete

XScorpion2 commented:
The corne basically calls straight into what that calls bypassing it if I remember correctly.

XScorpion2 commented:
i2c split code is in a much better state, however timer sync is not working out as I was hoping. Will implement a new synced_timer that wraps timer in non split keyboard mode so systems that need to be timer driven from the master half can just use that without needing to switch timers based on compile flags.

drashna commented:
Well, typing works over i2c now!
But rgblight doesn't.  Looks like the syncing for it doesn't work.

XScorpion2 commented:
Ok cool, just need to resolve the sync timer then.

Lenbok commented:
Just as a data point, I have converted my crkbd to use this PR plus #6001 and have switched my i2c redox (just using rgblight, not rgb_matrix) onto this as well for testing, can confirm that the rgblight syncing is not working. Happy to help test updates.

drashna commented:
Serial works 100%, but i2c has issues.   But only with animations.  If it's solid color, it works just fine. So it's definitely an issue with timer syncing. 

Though, it's getting there! 

@Lenbok awesome!

XScorpion2 commented:
Ok, if it is just i2c then that narrows down the issue. should be fixable tonight

XScorpion2 commented:
I2C sync timer is fixed

Lenbok commented:
```
Compiling: quantum/split_common/transport.c                                                        quantum/split_common/transport.c: In function ‘transport_master’:
quantum/split_common/transport.c:89:56: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
   i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_SYNC_TIME_START, (void *)sync_time, sizeof(i2c_buffer->sync_time), TIMEOUT);
                                                        ^
cc1: all warnings being treated as errors
 [ERRORS]
 |
 |
 |

```

XScorpion2 commented:
@Lenbok forgot the ref, pushed a fix

Lenbok commented:
Compiles now, but the slave side is not syncing properly. e.g. for breathing mode, after a short period (sometimes <1s, sometimes several seconds), the slave starts flashing different colors (like it's changing hsv values). For moving modes like knight/snake, the slave side one goes massively fast. Then it might come right for a few seconds, then speed up again.

XScorpion2 commented:
Hmm, ok might need more rgblight work then... Can you try rgb matrix?

XScorpion2 commented:
I reverted the sync timer changes to rgblight as I'm not sure if they are even necessary.

drashna commented:
that seems to work.

Lenbok commented:
Yep, the rgblight redox is working fine for me now too.

I don't have my crkbd here (it's at work), so to test the rgb_matrix I hacked up a config for my redox. Compiled and flashed the same hex to each side. Both sides are running the animations, but they are not in the same mode?!? When I press RGB_MOD both sides advance through the modes (so they are consistently not in the same mode). 

Video: https://photos.app.goo.gl/HmqYaxBvaGd6qnKY9
My test branch: https://github.com/Lenbok/qmk_firmware/commit/0cd7037d927b057fbf022fdfd7962e3ce15542ee

XScorpion2 commented:
@Lenbok ya the halves can get out of sync as there isn't any actual state sync for rgb matrix. I'm relying on them staying in sync by ensuring what keycodes happen on master also happen on the slave. I suspect your slave didn't have the same initial eeprom state as the master causing the two sides to be incorrect. To sync them back up add a custom keycode that you can hit to reset the rgb eeprom to default:

```
        case RGBRST:
            {
                if (record->event.pressed)
                    eeconfig_update_rgb_matrix_default();
            }
            return false;
```

Lenbok commented:
Can you do something along the lines of what rgblight does to sync across changes to the config? At startup and whenever the config changes it syncs across. The slave is basically running off the version it gets from the master and doesn't care about (or modify) what's in it's eeprom config.

drashna commented:
also, `EEP_RST` runs on both halves.  I think it should stay that way.

And I agree with @Lenbok, we should sync the rgbmatrix config, the say way that the rgblight config is.  Well, not exactly. It should be synced better, since that has bugs.

XScorpion2 commented:
Ya, I plan on adding something, just need some free time.

drashna commented:
One thing I've noticed recently is that if only one side of my crkbd is plugged in, the matrix does not render, at all

XScorpion2 commented:
hmm, odd, will check where I dropped the sync timer update logic, it might be after the is connected code and should be moved before possibly.
Just rebased on latest master, so you might want to try the branch again to see if it wasn't fixed already.

drashna commented on quantum/split_common/transport.c:
```diff
@@ -228,15 +244,22 @@ bool transport_master(matrix_row_t matrix[]) {
     return true;
 }
 
 void transport_slave(matrix_row_t matrix[]) {
```
Looks like this got messed up somehow. 

```suggestion
void transport_slave(matrix_row_t master_matrix[], matrix_row_t slave_matrix[] {
```

XScorpion2 replied:
good catch

drashna commented:
And yeah, tested, still has the issue where it stop *any* changes to the rgb matrix if the second half isn't attached.

XScorpion2 commented:
@drashna pushed a fix for that, also you might want to try enabling SERIAL_USE_MULTI_TRANSACTION that will reduce how much it attempts to sync each transport call and should help with deadlocks. At least I think, this is still very much not my area of knowledge.

drashna commented:
Yup, that change fixes it

Lenbok commented on quantum/split_common/matrix.c:
```diff
@@ -331,11 +331,8 @@ uint8_t matrix_scan(void) {
-#ifdef ENCODER_ENABLE
-        encoder_read();
-#endif
-        matrix_slave_scan_user();
```
It looks like that removes the only reference to `matrix_slave_scan_user`, nowhere else calls it, so it can be removed entirely now?

XScorpion2 replied:
Ya, now that this feature has stabilized I'll do a cleanup pass on everything

XScorpion2 replied:
Added matrix_slave_scan_user back in for when mirroring is not used

drashna commented on quantum/split_common/transport.c:
```diff
@@ -79,12 +84,18 @@ bool transport_master(matrix_row_t matrix[]) {
 }
 
-void transport_slave(matrix_row_t matrix[]) {
+void transport_slave(matrix_row_t master_matrix[], matrix_row_t slave_matrix[] {
```
```suggestion
void transport_slave(matrix_row_t master_matrix[], matrix_row_t slave_matrix[]) {
```

drashna commented:
Actually, it looks like that we could use the `SPLIT_TRANSPORT` option here, and specify a different file even. Or just add a define (eg `SPLIT_TRANSPORT = mirrored`)

drashna commented on quantum/quantum.c:
```diff
@@ -277,6 +277,9 @@ bool process_record_quantum(keyrecord_t *record) {
     switch (keycode) {
         case RESET:
             if (record->event.pressed) {
+#ifdef SPLIT_KEYBOARD
```
```suggestion
#ifdef SPLIT_TRANSPORT_MIRROR
```

drashna commented on quantum/split_common/transport.c:
```diff
@@ -123,6 +134,10 @@ typedef struct _Serial_s2m_buffer_t {
 } Serial_s2m_buffer_t;
 
 typedef struct _Serial_m2s_buffer_t {
+    matrix_row_t mmatrix[ROWS_PER_HAND];
```
```suggestion
#ifdef SPLIT_TRANSPORT_MIRROR
    matrix_row_t mmatrix[ROWS_PER_HAND];
#endif
```

drashna commented on quantum/split_common/transport.c:
```diff
@@ -216,7 +232,8 @@ bool transport_master(matrix_row_t matrix[]) {
     for (int i = 0; i < ROWS_PER_HAND; ++i) {
-        matrix[i] = serial_s2m_buffer.smatrix[i];
+        slave_matrix[i] = serial_s2m_buffer.smatrix[i];
+        serial_m2s_buffer.mmatrix[i] = master_matrix[i];
```
```suggestion
#ifdef SPLIT_TRANSPORT_MIRROR
        serial_m2s_buffer.mmatrix[i] = master_matrix[i];
#endif
```

drashna commented on quantum/split_common/transport.c:
```diff
@@ -228,15 +245,22 @@ bool transport_master(matrix_row_t matrix[]) {
     for (int i = 0; i < ROWS_PER_HAND; ++i) {
-        serial_s2m_buffer.smatrix[i] = matrix[i];
+        serial_s2m_buffer.smatrix[i] = slave_matrix[i];
+        master_matrix[i] = serial_m2s_buffer.mmatrix[i];
```
```suggestion
#ifdef SPLIT_TRANSPORT_MIRROR
        master_matrix[i] = serial_m2s_buffer.mmatrix[i];
#endif
```

XScorpion2 commented:
@drashna agreed, that looks like the minimal change and a great idea. I'll poke more at it this weekend.

XScorpion2 commented:
@drashna So far it looks like it's working. Could use your eyes on this as well.

joric commented:
It would be great to have an ability to redefine rgb matrix for the selected layout. E.g. in crkbd it's in rev1.c and can't be excluded/redefined. See https://github.com/joric/qmk/wiki/jorne#rgb-matrix

drashna commented:
@XScorpion2 I've been testing with these changes.  With the corne, you need to disable the "isLeftHnad" modification of the matrix, so ... that's a plus.

@joric I'm not sure what you mean.  the led config can be changed at runtime, and in fact, we're doing that for the corne, currently.  

However, yeah, you can't change it based on layout, since there is no way to detect that.  But you could reconfigure it in the keymap.

joric commented:
@drashna I just want that different keymaps could use different g_led_config, it's impossible for now because g_led_config is defined in rev1.c and rev1,c can't be replaced or excluded by the keymap configuration, because rev1.c is global.

drashna commented:
Well, I've been testing both mirrored and not mirrored, to ensure that existing behavior is preserved.   This doesn't. It instead scans on the slave side still.   You can verify this with the OLED keylogger code that I have.  `process_record_user` is still getting called. 

However, I have messed with things, locally, and do have it working for me.

XScorpion2 commented:
@drashna keyboard.c is fixed. my fault.

Lenbok commented:
@XScorpion2 Would you expect that both sides of the keyboard should be running the same mode now? I flashed the latest version of this pr with the corne split conversion to my corne, and the two sides started up in different modes (and RGB_MOD stepped each side, so they stayed in different modes).

An EEP_RST got both sides back to the same mode, but I would hope that this isn't required in general.

XScorpion2 commented:
@Lenbok I haven't followed back up on this yet. The plan is to do some minimal state sync between the two halves but have not had a chance to get it implemented.

drashna commented:
There are some merge conflicts here.  I can resolve them, if needed.

ShadowProgr commented on common_features.mk:
```diff
@@ -397,6 +397,10 @@ ifeq ($(strip $(SPLIT_KEYBOARD)), yes)
         # Unused functions are pruned away, which is why we can add multiple drivers here without bloat.
+        ifeq ($(strip $(SPLIT_TRANSPORT)), mirror)
+            OPT_DEFS += -DSPLIT_TRANSPORT_MIRROR
+        endef 
```
Typo here
```suggestion
        endif 
```

drashna commented:
Sorry if I have overstepped, but I've updated a couple things (repeatedly now). 

One thing that has been mentioned is that thi is missing the "get_slave_status" stuff, so that the matrix is only updated if there is an actual change.


#7732 is an implementation of that, but it would be nice to get in here.  Though ... I am not sure if that wil help get this merged.

XScorpion2 commented:
@drashna _GASP how dare you sir!_
No seriously you are fine. Saw #9001 and thought that would be a nice addition to this pr. still need to add a rgb config sync to prevent halves from possibly getting out of sync.

Gansgar commented:
I've now successfully merged the current master with this branch and would be ready to implement the state sync. I've actually done that before, got stuck at the timer sync and only later realized, that this pull request exists. As such I've the code from that approach still in my git repo and would be happy to implement it into this one.

Gansgar commented:
I've created a new merge request with my changes. See #9613

XScorpion2 commented:
Closed for the updated version https://github.com/qmk/qmk_firmware/pull/11055
